src/restic/snapshot_filter.go

```diff
@@ -124,84 +77,16 @@ func y(d time.Time) int {
 	return d.Year()
 }
 
-// apply moves snapshots from Unprocess to either Keep or Remove. It sorts the
-// snapshots into buckets according to the return value of fn, and then moves
-// the newest snapshot in each bucket to Keep and all others to Remove. When
-// max snapshots were found, processing stops.
-func (f *filter) apply(fn func(time.Time) int, max int) {
-	if max == 0 || len(f.Unprocessed) == 0 {
-		return
-	}
-
-	sameBucket := Snapshots{}
-	lastBucket := fn(f.Unprocessed[0].Time)
-
-	for len(f.Unprocessed) > 0 {
-		cur := f.Unprocessed[0]
-
-		bucket := fn(cur.Time)
-
-		// if the snapshots are from a new bucket, forget all but the first
-		// (=last in time) snapshot from the previous bucket.
-		if bucket != lastBucket {
-			f.Keep = append(f.Keep, sameBucket[0])
-			f.Remove = append(f.Remove, sameBucket[1:]...)
-
-			sameBucket = Snapshots{}
-			lastBucket = bucket
-			max--
-
-			if max == 0 {
-				return
-			}
-		}
-
-		// collect all snapshots for the current bucket
-		sameBucket = append(sameBucket, cur)
-		f.Unprocessed = f.Unprocessed[1:]
-	}
-
-	// if we have leftovers, process them too.
-	if len(sameBucket) > 0 {
-		f.Keep = append(f.Keep, sameBucket[0])
-		f.Remove = append(f.Remove, sameBucket[1:]...)
-	}
-}
-
-// keepTags marks the snapshots which have all tags as to be kept.
-func (f *filter) keepTags(tags []string) {
-	if len(tags) == 0 {
-		return
-	}
+var a int
 
-	unprocessed := f.Unprocessed[:0]
-	for _, sn := range f.Unprocessed {
-		if sn.HasTags(tags) {
-			f.Keep = append(f.Keep, sn)
-			continue
-		}
-		unprocessed = append(unprocessed, sn)
-	}
-	f.Unprocessed = unprocessed
+// always retuns a unique number for d.
+func always(d time.Time) int {
+	a++
+	return a
```

Comment:
Since we have nanosecond precision in the snapshot timestamps, I think we can just use `int(d.UnixNano())` here. That should be unique, always (even if the `int64` is casted to an `int32`). This removes the need for the singleton `a`.